From 0afeb33e4ed03ed689c4d444b5ded7950cb7ea02 Mon Sep 17 00:00:00 2001
From: Martin Fick <mfick@codeaurora.org>
Date: Fri, 10 Jun 2011 14:43:52 -0600
Subject: [PATCH] Add SideBySide and Unified links to Differences top level menus

These new menu entries allow a user to switch view types easily
without returning to the ChangeScreen.  Also, they double as a
way to hide the header on the PatchScreen (when clicking on the
currently displayed type) and thus this patch also eliminates
the 'Diff Only' link.

Change-Id: I769638c133d7b3e17bed640c8080ea2b2e685a1b
---
 .../java/com/google/gerrit/client/Dispatcher.java  |    8 ++++++
 .../main/java/com/google/gerrit/client/Gerrit.java |   25 +++++++++++++++++++-
 .../com/google/gerrit/client/GerritConstants.java  |    1 -
 .../gerrit/client/GerritConstants.properties       |    1 -
 .../google/gerrit/client/patches/PatchScreen.java  |   10 +++++--
 5 files changed, 39 insertions(+), 6 deletions(-)

diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/Dispatcher.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/Dispatcher.java
index 8b82a30..7e9ffc9 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/Dispatcher.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/Dispatcher.java
@@ -83,6 +83,14 @@ public class Dispatcher {
     return "patch," + type + "," + id.toString();
   }
 
+  public static String toPatch(final PatchScreen.Type type, final Patch.Key id) {
+    if (type == PatchScreen.Type.SIDE_BY_SIDE) {
+      return toPatchSideBySide(id);
+    } else {
+      return toPatchUnified(id);
+    }
+  }
+
   public static String toAccountGroup(final AccountGroup.Id id) {
     return "admin,group," + id.toString();
   }
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/Gerrit.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/Gerrit.java
index 021c70b..2dcac32 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/Gerrit.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/Gerrit.java
@@ -16,6 +16,7 @@ package com.google.gerrit.client;
 
 import com.google.gerrit.client.auth.openid.OpenIdSignInDialog;
 import com.google.gerrit.client.auth.userpass.UserPassSignInDialog;
+import com.google.gerrit.client.changes.ChangeConstants;
 import com.google.gerrit.client.changes.ChangeListScreen;
 import com.google.gerrit.client.patches.PatchScreen;
 import com.google.gerrit.client.rpc.GerritCallback;
@@ -33,6 +34,7 @@ import com.google.gerrit.reviewdb.Account;
 import com.google.gerrit.reviewdb.AccountDiffPreference;
 import com.google.gerrit.reviewdb.AccountGeneralPreferences;
 import com.google.gerrit.reviewdb.AuthType;
+import com.google.gerrit.reviewdb.Patch;
 import com.google.gwt.core.client.EntryPoint;
 import com.google.gwt.core.client.GWT;
 import com.google.gwt.core.client.JavaScriptObject;
@@ -66,6 +68,7 @@ import java.util.ArrayList;
 
 public class Gerrit implements EntryPoint {
   public static final GerritConstants C = GWT.create(GerritConstants.class);
+  public static final ChangeConstants CC = GWT.create(ChangeConstants.class);
   public static final GerritMessages M = GWT.create(GerritMessages.class);
   public static final GerritResources RESOURCES =
       GWT.create(GerritResources.class);
@@ -491,7 +494,8 @@ public class Gerrit implements EntryPoint {
     patchScreen = null;
     diffBar = new LinkMenuBar();
     menuLeft.addInvisible(diffBar, C.menuDiff());
-    addDiffLink(diffBar, C.menuDiffMain(), PatchScreen.TopView.MAIN);
+    addDiffLink(diffBar, CC.patchTableDiffSideBySide(), PatchScreen.Type.SIDE_BY_SIDE);
+    addDiffLink(diffBar, CC.patchTableDiffUnified(), PatchScreen.Type.UNIFIED);
     addDiffLink(diffBar, C.menuDiffCommit(), PatchScreen.TopView.COMMIT);
     addDiffLink(diffBar, C.menuDiffPreferences(), PatchScreen.TopView.PREFERENCES);
     addDiffLink(diffBar, C.menuDiffPatchSets(), PatchScreen.TopView.PATCH_SETS);
@@ -605,6 +609,25 @@ public class Gerrit implements EntryPoint {
       });
   }
 
+  private static void addDiffLink(final LinkMenuBar m, final String text,
+      final PatchScreen.Type type) {
+    m.addItem(new LinkMenuItem(text, "") {
+        @Override
+        public void go() {
+          if (patchScreen != null) {
+            if (type == patchScreen.getPatchScreenType()) {
+              patchScreen.setTopView(PatchScreen.TopView.MAIN);
+              AnchorElement.as(getElement()).blur();
+            } else {
+              final Patch.Key id = patchScreen.getPatchKey();
+              setTargetHistoryToken(Dispatcher.toPatch(type, id));
+              super.go();
+            }
+          }
+        }
+      });
+  }
+
   private static void addDocLink(final LinkMenuBar m, final String text,
       final String href) {
     final Anchor atag = anchor(text, "Documentation/" + href);
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/GerritConstants.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/GerritConstants.java
index 34f250c..b4293dc 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/GerritConstants.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/GerritConstants.java
@@ -62,7 +62,6 @@ public interface GerritConstants extends Constants {
   String menuMyStarredChanges();
 
   String menuDiff();
-  String menuDiffMain();
   String menuDiffCommit();
   String menuDiffPreferences();
   String menuDiffPatchSets();
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/GerritConstants.properties b/gerrit-gwtui/src/main/java/com/google/gerrit/client/GerritConstants.properties
index 2965d6c..7b06788 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/GerritConstants.properties
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/GerritConstants.properties
@@ -45,7 +45,6 @@ menuMyStarredChanges = Starred Changes
 menuMyWatchedChanges = Watched Changes
 
 menuDiff = Differences
-menuDiffMain = Diff Only
 menuDiffCommit = Commit Message
 menuDiffPreferences = Preferences
 menuDiffPatchSets = Patch Sets
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java
index 830e1ba..1adadbf 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java
@@ -70,7 +70,7 @@ public abstract class PatchScreen extends Screen implements
     }
 
     @Override
-    protected PatchScreen.Type getPatchScreenType() {
+    public PatchScreen.Type getPatchScreenType() {
       return PatchScreen.Type.SIDE_BY_SIDE;
     }
   }
@@ -87,7 +87,7 @@ public abstract class PatchScreen extends Screen implements
     }
 
     @Override
-    protected PatchScreen.Type getPatchScreenType() {
+    public PatchScreen.Type getPatchScreenType() {
       return PatchScreen.Type.UNIFIED;
     }
   }
@@ -352,7 +352,11 @@ public abstract class PatchScreen extends Screen implements
 
   protected abstract AbstractPatchContentTable createContentTable();
 
-  protected abstract PatchScreen.Type getPatchScreenType();
+  public abstract PatchScreen.Type getPatchScreenType();
+
+  public Patch.Key getPatchKey() {
+    return patchKey;
+  }
 
   protected void refresh(final boolean isFirst) {
     final int rpcseq = ++rpcSequence;
-- 
1.7.4.1

